### ISSUE TITLE
Total size of L2 cache size

### ISSUE THREAD

**Issue by xmrig** [anrieff/libcpuid#97]
How detect total size of L2 cache size?
For example: i7 6700 `cpu_id_t::l2_cache` is 256 so 256 * 4 = 1 MB it's right.

but `AMD Athlon(tm) X4 860K Quad Core Processor` 4 MB of L2 cache.
Thank you.

**Comment by anrieff** [comment]
Hi!
Thanks for your report.
Ideally, you should be able to compute that by `cpu_id_t::num_cores` × `cpu_id_t::l2_cache`. As you note, on i7 6700 you'll get the correct number; on the 860K you'd get 4*2MB = 8MB which is incorrect. The culprit is that on Bulldozer and later AMD CPUs, `cpud_it_t::num_cores` is probably misleading and should be half of that number, very much like the Intel's hyperthreading. The problem is that AMD's CMT and Inte's SMT are not identical so this is a controversial question.

E.g., you may ask how many cores does the 860K really have?
- The brand string says it's a quad-core processor
- In integer operations it behaves as a quad-core processor
- In floating-point operations it behaves as a dual-core processor (very roughly)
- Cache-wise it is a dial-core processor

Then, for consistency I feel now it's probably best to consider 860K a 2-core processor with hyper-threading. @X0rg, what do you think on this? And would CPU-X be severely affected after the "downgrade"?

**Comment by xmrig** [comment]
Maybe it good workaround for me?
If CPU vendor is AMD and no L3 cache and `l2_assoc == 16` so CPU maybe contains shared L2 cache.

**Comment by anrieff** [comment]
This might work, but seems too hacky to me.
If you want a quick workaround, probably easier one will be to check `cpu_id_t::ext_family >= 21`. I think it may be safe to assume, that if it's 21 or more, then you have CMT, so the L2 cache is shared.

**Comment by X0rg** [comment]
According to [AMD](http://products.amd.com/en-us/search/CPU/AMD-Athlon%E2%84%A2/AMD-Athlon%E2%84%A2-X4/860K-with-New-Thermal-Solution/133), Athlon X4 860K is a 4 core with 4 threads.
Before Zen, the SMT feature was not. So, considering this CPU as a 2 core with HT seems wrong: they are modules in fact (2 modules = 4 cores), but there isn't SMT on Kaveri.

I'm without opinion. Don't worry about CPU-X; do what is the best for libcpuid. Thank you for asking.

**Comment by phillipsjk** [comment]
I was able to reproduce this under WinXP-64, on an Intel Core 2 Duo E8400, with xmrig.

That part has a 6MB "Smart Cache" shared between the cores.

Presumably because each core is able to use the whole cache, it gets detected as 12MB of L2 cache.

**Comment by X0rg** [comment]
This issue is being addressed in #168. The idea is, assuming the total size of Ln cache size is needed, `cpu_id_t::ln_instances` × `cpu_id_t::ln_cache` formula can be used.

In case of hybrid CPU, `system_id_t::ln_total_instances` provides the grant total of cache instances. For example, following code can be used to display L3 cache size:
```c
int32_t total_l3_size = 0;
struct system_id_t system;
cpu_identify_all(NULL, &system);

if ((system.l3_total_instances == 1) && (system.num_cpu_types > 1)) {
    printf("L3 size (shared between all types of core) is %i\n", system.cpu_types[0].l3_cache)
}
else {
    for (uint8_t cpu_type_index = 0; cpu_type_index < system.num_cpu_types; cpu_type_index++) {
        printf("L3 size for type %u is %i×%i\n", cpu_type_index, system.cpu_types[cpu_type_index].l3_instances, system.cpu_types[cpu_type_index].l3_cache);
        total_l3_size += system.cpu_types[cpu_type_index].l3_instances * system.cpu_types[cpu_type_index].l3_cache;
    }
    printf("Total L3 size is %i\n", total_l3_size);
}
```
For Intel Alder Lake CPUs, there are 2 types of CPU core (P-cores and E-cores), but all of them share the same L3 cache on the same die.
However, for AMD Zen 1 and Zen 2 CPUs, there are 2 CCX per CCD, i.e. 2 instances of L3 cache per die.
